### PR TITLE
Dependency generator optimization and cleanup

### PR DESCRIPTION
**Pull request by pmatilai**
This basically turns the dependency generation order inside out: previously we ran through the file list one by one, running dependency generators for all types the file was found to match, ie in file:attr:deptype order. Now we do the exact opposite: deptype:attr:file, which allows us to optimize to only expand all those macros once per deptype / attr instead of once per file. 

Sadly all that is lost in the noise of actually forking + executing those dependency generators, but this is basically a prerequisite for the far more important optimizations, such as (some day) teaching the generators to work on multiple files at once and ultimately, multiple generators in parallel.

### CODE REVIEW

File: build/rpmfc.c [hunk 1]
```diff
@@ -576,19 +569,16 @@ static void exclFini(struct exclreg_s *excl)
...
- * @param depname	"provides" or "requires"
  * @param dsContext	RPMSENSE_FIND_PROVIDES or RPMSENSE_FIND_REQUIRES
  * @param tagN		RPMTAG_PROVIDENAME or RPMTAG_REQUIRENAME
  * @return		0 on success
  */
 static int rpmfcHelper(rpmfc fc, int ix, const struct exclreg_s *excl,
-		       const char *nsdep, const char *depname,
-		       rpmsenseFlags dsContext, rpmTagVal tagN)
+		       rpmsenseFlags dsContext, rpmTagVal tagN,
+		       const char *namespace, const char *cmd)
```
Comment (ffesti) [inline build/rpmfc.c]: The new params are not added to the doc string

Comment (pmatilai) [reply build/rpmfc.c]: Urgh, yeah. Mind if I just nuke the doc string, it's just an internal helper after all...?

Comment (ffesti) [reply build/rpmfc.c]: Fine with me.

Comment (pmatilai) [reply build/rpmfc.c]: Okay, axed them all in a separate commit.

**Comment by ffesti** [comment]
I was first concerned that all this "putting stuff in data structures" might create overhead. After having a deeper look this is not a problem. Storing a few more integers  while we have the whole set of file names in memory will probably not kill anyone.

**Comment by ffesti** [comment]
Looks good. Please squash the doc string fix and push upstream!

**Comment by pmatilai** [comment]
Oh, I left the doc string fix as a separate commit because it eliminates *all* the useless docstrings from rpmfc so its not really related to these.

**Comment by pmatilai** [comment]
And yeah there is going to be some overhead from using a hash, but otoh this also eliminates a whole lot of overhead that is currently running on each file, and if we ever want to run more than one file at a time through a generator there's not a whole lot of choice.

**Comment by pmatilai** [comment]
Anyway, dropped the docstring change from the "expand all macros" commit, its all in the cleanup commit instead.

**Pull request merged**